### DISCUSSION TITLE
Why PrimeTween doesn't use extension methods?

### DISCUSSION THREAD

**Discussion by KyryloKuzyk** [KyryloKuzyk/PrimeTween#3] [Announcements]
Why PrimeTween doesn't use extension methods?
---

### What's an extension method, anyway?
Extension methods in C# is a syntactic sugar feature that allows one to 'add' new functionality to other classes without modifying their source code.  
In reality, an extension method is just a static method that accepts the type you're 'extending' as a first parameter. But for the user of your extension method, it looks like a regular method of that type. Cool, huh? But are there any downsides?

I'll review the potential downsides of extension methods on the example of DOTween, but this article can be applied to extension methods in general.

### Extension methods deceive.
The first downside comes directly from the advantage - extension methods 'pretend' that an object has the functionality when in reality it doesn't. When you type `transform.`, your IDE will suggest all third-party extension methods alongside built-in Unity methods. You may think you're using a native Unity method when in reality you may be using a third-party dependency.  

Let's assume you're making a shooter game and you see this code. Can you tell at the first glance what the code will do?
```csharp
enemy.DOKill();
```
Will the code kill the enemy? Will it destroy its GameObject? Or maybe the enemy will kill the player? It turns out, it's a DOTween's extension method that will stop all tweens and delays running on the enemy's MonoBehaviour. Unexcepted, right?  

With PrimeTween, you can always be confident in what the code actually does:
```csharp
Tween.StopAll(onTarget: enemy);
```

### Extension methods hide complexity.
Let's consider an example:
```csharp
transform.Translate(Vector3.one);  
transform.DOMove(Vector3.one, duration: 1);
```

Which line is more 'significant'? Or simply put, which method call do 'more stuff'? At first glance, both method calls have the same significance, they look similar and they should move the transform in some way.  
In reality, `Translate()` will immediately move the object, while `DOMove()`  does a lot of things under the hood: it will create an animation that will change the position every frame for the next second. The look of the `DOMove()` method doesn't convey its complexity in comparison to `Translate()`.

With PrimeTween, the `Tween.Position()` clearly indicates we're using an animation library:
```csharp
transform.Translate(Vector3.one);  
Tween.Position(transform, Vector3.one, duration: 1);
```

### Extension methods can hurt readability.
Let's consider how extension methods can impact readability. The next piece of code reads naturally, like this: _animate position from 10 to 20 in 1 second_.
```csharp
Tween.PositionY(transform, startValue: 10, endValue: 20, duration: 1);
```

While this reads awkwardly: _animate position to 20 in 1 second from 10_.
```csharp
transform.DOMoveY(endValue: 20, duration: 1).From(fromValue: 10);
```

### Builder Pattern in disguise.
A lot of times, extension methods are used as a Builder pattern. Builder pattern has a lot of criticism on its own, but I will touch only on the aspect of how it affects API design.  
In DOTween, when you type `tween.` in your IDE, you'll see a bunch of different methods that can be split into two different categories: 1) **creation** methods and 2) **manipulation** methods. All of these methods are in the same bucket, and only by studying the documentation and examples, you can understand to which group every one of them belongs.  
Here is an example. All of these methods don't make sense after the tween has started, they serve only as **creation** methods to start a tween:
```csharp
transform.DOMove(targetPos, duration)
  .SetEase(Ease.OutQuad)
  .SetLoops(cycles, LoopType.Yoyo)
  .SetDelay(delay)
  .SetUpdate(useUnscaledTime)
  .SetLink(transform.gameObject)
  .Play();
```
In PrimeTween, all the above code is as simple as passing parameters to a method. And when you type `tween.`, you'll see only a small subset of **manipulation** methods.
```csharp
Tween.Position(transform, targetPos, duration, Ease.OutQuad, cycles, CycleMode.Yoyo, delay, 0, useUnscaledTime);
```
Or you can use TweenSettings if you don't like a long chain of parameters or wish to improve readability. You can also pass the TweenSettings structure to other methods, serialise it in Inspector, or share it between different tween types.
```csharp
Tween.Position(transform, targetPos, new TweenSettings {
    duration = 1f,
    ease = Ease.OutQuad,
    cycles = 2,
    cycleMode = CycleMode.Yoyo,
    startDelay = 0.1f,
    useUnscaledTime = true,
    useFixedUpdate = true
});
```

### Unity doesn't use extension methods.
I wanted to make PrimeTween feel 'native' in Unity. I've never encountered an extension method in UnityEngine or any official UPM package. For example, `Mathf` class or `Vector3.Dot/Cross/Distance()` could've been implemented as extension methods, but they are not.
Or `Physics.Raycast()` could've been an extension method. Imagine how 'cool' it would be to write something like this:
```csharp
camera.ViewportPointToRay(screenPoint).Raycast().SetDistance(10).SetMask(layerMask).Execute();
```

But this is not the case. Unity uses good-old static `Physics.Raycast()` method that simply accepts parameters. And there is a reason for it - Unity is building a tool for millions of developers and not a project to show cool syntactic sugar tricks.

### Performance impact.
I put the performance in last place because users' code readability and maintainability are far more important in my opinion. The performance of extension methods in themselves is never a problem, because they are essentially static methods. But when extension methods are used as the Builder Pattern to apply parameters, this has a measurable impact on performance.

For example, in DOTween, starting an animation with all parameters adds an extra 4.47 ms (per 100,000 iterations) in comparison to the simplest animation. While in PrimeTween, passing all parameters to a static method adds just [0.96 ms](https://github.com/KyryloKuzyk/PrimeTween/discussions/8). It will unlikely be noticeable in a real project, but it's a nice bonus to all previous benefits.

### "But I still miss DOTween's extension methods".
We've been all using DOTween for many years and have an addiction to its extension methods, I get that.  

PrimeTween comes with a simple [interface adapter](https://github.com/KyryloKuzyk/PrimeTween#dotween-adapter) so you can use your favorite extension methods and get all the performance benefits of PrimeTween. PrimeTween doesn't conflict with DOTween and they can be used side-by-side in one project. After enabling the adapter, just type `using PrimeTween;` instead of `using DG.Tweening;` at the top of your script and you're good to go.

### Conclusion.
Should you ditch extension methods altogether? Absolutely not. For example, LINQ is a good example of extension methods used correctly.

But when you write your own custom extension methods, consider how much value they add, taking into account all aforementioned readability costs. Typically, you write the code once, then you read it hundreds of times while you work on a project. So when in doubt, prefer code readability over the number of characters you have to type or how 'cool' the code may look.

Have something to add? Or disagree? You're welcome in the comments. Help me make PrimeTween better for the community.

**Comment by STARasGAMES** [comment]
With some things I completely agree, but I think some aspects are misrepresented. 


You have an example:
```c#
enemy.DOKill();
```
For a person seeing this piece of code for the first time and having no previous experience with DOTween I agree that it is misleading. 
But, it's not just ```enemy.Kill();```, it's **DO**Kill, which for the entire codebase consistently indicates that this method is part of **DO**Tween library. Developer needs to understand this concept only once, it's quite simple afterward. 

Then you show this:
```c#
Tween.StopAll(onTarget: enemy);
```
This is somewhat better than DOTween in terms of naming, but you did a little manipulation - no one writes argument names. In the real world this would look like this:
```c#
Tween.StopAll(enemy);
```
And this is also not good at describing what this method does. Why `All` is used for `enemy` which is a singular form? After `stopping` is there a way to continue or tweens are completely erased? Yes, the term 'Kill' is not widely used, but it feels like better describing what will happen with tweens.


My take on this:
```c#
PrimeTween.StopAllOnTarget(enemy); // or KillAllOnTarget
```
Or even like this:
```c#
enemy.AsTweenable().StopAllTweens(); // or KillAllTwins... just having fun at this point :)
```

### Extension methods can hurt readability.
```c#
Tween.PositionY(transform, startValue: 10, endValue: 20, duration: 1);
```
I can't remember the single time I needed to specify the startValue, that's why I didn't even know about existence of `.From()`. Just can't relate to this example. It feels like a super edge case. Could there be another example?


### Builder Pattern in disguise.
```c#
transform.DOMove(targetPos, duration)
  .SetEase(Ease.OutQuad)
  .SetLoops(cycles, LoopType.Yoyo)
  .SetDelay(delay)
  .SetUpdate(useUnscaledTime)
  .SetLink(transform.gameObject)
  .Play();
```
The `Set` is not the right word indeed. It should be `With`, cuz `Set` feels like I can set value at any time, while in reality, as you said, I can't set value after tween is started.
<br>
Your solution:
```c#
Tween.Position(transform, targetPos, duration, Ease.OutQuad, cycles, CycleMode.Yoyo, delay, 0, useUnscaledTime);
```
The above code looks pretty and convenient when you have properly named variables.
But my assumption is that in most cases it looks like this:
```c#
Tween.Position(transform, new Vector3(0,1,0), 0.2f, Ease.OutQuad, 4, CycleMode.Yoyo, 0.3f, 0, true);
```
Wish you luck in understanding this mess XD
Now seriously, these magic values are bad and developers should define properly named variables, but let's face it - only a few of us do it.

Now look at this:
```c#
transform.DOMove(new Vector3(0,1,0), 0.2f)
  .WithEase(Ease.OutQuad)
  .WithLoop(4, LoopType.Yoyo)
  .WithDelay(0.3f)
  .WithUnscaledTime(true)
  .WithLink(transform.gameObject)
  .Play();
```
This approach is easier to read when magic values are used. And it's super easy to write and iterate, there is no requirement on argument order, and you can add a new "argument" anywhere you want in that chain. Code completion is also easier, all you have to do is type `.With` and you have a list of all available options, which even could be extended with your or your teammate's custom extensions!

And a bit of my personal preference(or should I say personal frustration) - methods with too many arguments. I use Unity for more than 10 years, and every time I type `Physics.Raycast()` I know that next few minutes I will struggling to find the proper overload for my particular case. Rider helps with this a lot lately, but it's still quite overwhelming.

**Reply by KyryloKuzyk** [reply]
Thank you for joining the discussion, very valid points!

> It's DOKill, which for the entire codebase consistently indicates that this method is part of DOTween library

This is a very valid argument. But it assumes you have experience with DOTween and you know the project very well. But if you're joining someone else's project or revisiting a project you haven't touched in a while, the chances are high that you'll be very confused with `enemy.DOKill()`.

My point here is more about '**how fast**' you can identify what method you're using. And I argue that it's a lot easier to understand that you're using an animation library when you see `Tween.Position(transform)` instead of `transform.DOMove()`.

> No one writes argument names

Me neither :)  
All modern IDEs have an '**Inlay Hints**' feature that shows argument names and I believe it is turned on by default. This is how the code looks in IDE even when there are no parameter names in the code:
<img width="1237" alt="image" src="https://github.com/KyryloKuzyk/PrimeTween/assets/13222270/7a03385a-c2b5-44e9-a757-d7859834f60f">
Although when you read the code on GitHub or in Unity's quick preview, I agree that not having IDE hints can worsen the readability.

> I can't remember the single time I needed to specify the startValue, that's why I didn't even know about existence of .From(). Just can't relate to this example. It feels like a super edge case. Could there be another example?

The From() extension is fairly widespread. And while I was searching its usages on GitHub, I found another good example of using SetRelative() that's also used very often:
```csharp
transform.DOScale(Vector3.one, duration: 0.5f).SetRelative(true);
// The above code can be rewritten in a more readable fashion as:
Tween.LocalScale(transform, endValue: transform.localScale + Vector3.one, duration: 0.5f);
```

> I use Unity for more than 10 years, and every time I type Physics.Raycast() I know that next few minutes I will struggling to find the proper overload for my particular case. Rider helps with this a lot lately, but it's still quite overwhelming.

Yes, Physics.Raycast() can be overwhelming, I agree.  
But in PrimeTween there are only 3 mandatory parameters (target, endValue, and duration), all other six parameters are optional. And while having six optional parameters seems like a lot, it's all there is: parameters communicate all available properties an animation can **possibly** have. No more, no less.  
And once you've learned how to use one animation method, consider you've learned them all.

> And it's super easy to write and iterate, there is no requirement on argument order, and you can add a new "argument" anywhere you want in that chain.

Optional arguments can also be used in any order, you just have to specify their name.

Saying all that, I totally understand the concern of changing how people used to animate things previously. Adding extension methods to PrimeTween is absolutely possible (the adapter shows that), so if there would be a lot of unsatisfied users, there is always an option of adding extension methods.  
Adding things is easy, but once you release something, there is no way back. And I want to keep the API simple and accessible.

**Comment by krvs** [comment]
Static methods are faster because they doesn't have null comparison under the hood

**Reply by KyryloKuzyk** [reply]
Yes, almost all DOTween's extension methods have a null check under the hood, which theoretically should influence the performance in a bad way. 

However I'm currently working on a blog post comparing the performance of PrimeTween and DOTween, and while PrimeTween is about 2X times faster than DOTween, the impact of not using extension methods (and extra null checks) is very hard to measure even in big stress tests. This can be explained that in comparison to the total complexity of a tween library, a couple of extra null checks don't make much difference. Will release the blog and all benchmarks' source code soon to support my claims with numbers.

So I still believe the main reason for not using extension methods is improved readability and simpler API.

**Reply by KyryloKuzyk** [reply]
So I actually [tested](https://github.com/KyryloKuzyk/PrimeTween/discussions/8) the impact of extension methods on performance and the difference can be measured. 

In DOTween, starting an animation will all parameters adds an extra 4.47 ms (per 100,000 iterations) in comparison to the simplest animation. While in PrimeTween, passing all parameters to a static method adds just 0.96 ms. It will unlikely be noticeable in a real project, but it's a nice bonus to all readability benefits.

**Comment by remarxk** [comment]
> `Tween.Position(transform, targetPos, duration, Ease.OutQuad, cycles, CycleMode.Yoyo, delay, 0, useUnscaledTime);`

Don't you think this is a long piece of code? I think that when a method is exposed with more than four variables, the readability and usability of the method is greatly reduced. In fact, in my opinion, dotween's builder mode is more readable and easier to use.

> Unity doesn't use extension methods.

This is wrong. In fact, I remember unity having a lot of classes that use extension methods. I can't remember which ones, but I'm pretty sure the VisualElement in uitoolkit does.



DOTween's extension method works well, but it is crucial that his methods maintain a uniform naming style, that is, the DO beginning. If you're afraid of the trickery of the extension method, I think you can also use a uniform naming style so that the first thing you see is that this is the PrimeTween method.

I share STARasGAMES 'sentiments for the most part, and I hope you'll reconsider the usability of extension methods.

**Reply by KyryloKuzyk** [reply]
> Don't you think this is a long piece of code?

@remarxk It is! But the version with extension methods is not any better in my opinion.
In any case, this example is somewhat extraordinary. In the vast majority of cases, it would look like this:
`Tween.Position(transform, targetPos, duration, Ease.OutQuad);`

And only sometimes there will be `cycles, cycleMode` OR `delay` OR `useUnscaledTime`. Having all of those is quite rare, and using extension methods vs a long list of parameters is a matter of choice. PrimeTween compensates the long list of params with the ability to tweak them all via [Inspector](https://github.com/KyryloKuzyk/PrimeTween#inspector-integration).

> This is wrong. In fact, I remember unity having a lot of classes that use extension methods. I can't remember which ones, but I'm pretty sure the VisualElement in uitoolkit does.

I would love to see an example. The only time I encountered a builder pattern in the official Unity packages was in the Performance Testing API, which is a quite niche package. And, of course, I ended up forgetting to call the `.Run()` method at the end of the method chain the first time I used it :)

> DOTween's extension method works well, but it is crucial that his methods maintain a uniform naming style, that is, the DO beginning. 

This is just one of many approaches you can take with a tweening library. In fact, the most popular JS tweening libs don't use the approach with prefixes, and their users are perfectly fine with it. 
While designing PrimeTween, I had a specific vision for an API design, and I believe it resulted in a very simple and clean API. I encourage you to give PrimeTween's syntax a chance because in my opinion such API style leads to a more readable and maintainable code in the majority of cases.

If you'll be still missing extension methods and DO prefixes, PrimeTween has an interface [adapter](https://github.com/KyryloKuzyk/PrimeTween#dotween-adapter). With the adapter, you can continue writing your code in DOTween's syntax, while having all the performance benefits of PrimeTweeen.

**Reply by KyryloKuzyk** [reply]
@remarxk Also, if you want to be more explicit with parameter names, it's possible to do the following. And now the code reads the same way as with extension methods:
```csharp
Tween.Position(transform, targetPos, new TweenSettings {
    duration = 1f,
    ease = Ease.OutQuad,
    cycles = 2,
    cycleMode = CycleMode.Yoyo,
    startDelay = 0.1f,
    useUnscaledTime = true,
    useFixedUpdate = true
});
```

You can pass the TweenSettings structure to other methods, serialize it in Inspector, and share it between different tweens. All of this is not possible with extension methods.
